lib/server-builder.js

```diff
@@ -0,0 +1,161 @@
+
+class ServerBuilder {
+  constructor (args = {}) {
+    this.serverPath = args.serverPath;
+    this.showGradleLog = args.showGradleLog;
+
+    const buildConfiguration = args.buildConfiguration || {};
+
+    const versionConfiguration = buildConfiguration.toolsVersions || {};
+    const serverVersions = {};
+    for (const key of VERSION_KEYS) {
+      if (versionConfiguration[key]) {
+        serverVersions[key] = versionConfiguration[key];
+      }
+    }
+    this.serverVersions = serverVersions;
+
+    this.additionalAppDependencies = buildConfiguration.additionalAppDependencies || [];
+  }
+
+  async build () {
+    if (this.serverVersions[GRADLE_VERSION_KEY]) {
+      await this.setGradleWrapperVersion(this.serverVersions[GRADLE_VERSION_KEY]);
+    }
+    await this.insertAdditionalAppDependencies(this.additionalAppDependencies);
+    await this.runBuildProcess();
+  }
+
+  getCommand () {
+    const cmd = system.isWindows() ? 'gradlew.bat' : './gradlew';
+    let args = VERSION_KEYS
+      .filter(key => key !== GRADLE_VERSION_KEY)
+      .map(key => {
+        const serverVersion = this.serverVersions[key];
+        const gradleProperty = `appium${key.charAt(0).toUpperCase()}${key.slice(1)}`;
+        return serverVersion ? `-P${gradleProperty}=${serverVersion}` : null;
+      })
+      .filter(Boolean);
+
+    args.push('assembleAndroidTest');
+
+    return {cmd, args};
+  }
+
+  async setGradleWrapperVersion (version) {
+    const propertiesPath = path.resolve(this.serverPath, 'gradle', 'wrapper', 'gradle-wrapper.properties');
+    const originalProperties = await fs.readFile(propertiesPath, 'utf8');
+    const newProperties = this.updateGradleDistUrl(originalProperties, version);
+    await fs.writeFile(propertiesPath, newProperties, 'utf8');
+  }
+
+  updateGradleDistUrl (propertiesContent, version) {
+    return propertiesContent.split('\n')
+      .map(function (line) {
+        if (line.startsWith(GRADLE_URL_PREFIX)) {
+          return `${GRADLE_URL_PREFIX}${GRADLE_URL_TEMPLATE.replace('VERSION', version)}`;
+        } else {
+          return line;
+        }
+      })
+      .join('\n');
+  }
+
+  async insertAdditionalAppDependencies (additionalAppDependencies) {
+    const buildPath = path.resolve(this.serverPath, 'app', 'build.gradle');
+    const originalConfiguration = await fs.readFile(buildPath, 'utf8');
+    const newConfiguration = this.updateDependencyLines(originalConfiguration, additionalAppDependencies);
+    await fs.writeFile(buildPath, newConfiguration, 'utf8');
+  }
+
+  updateDependencyLines (configurationContent, additionalAppDependencies) {
+    if (!Array.isArray(additionalAppDependencies)) {
+      throw new Error('additionalAppDependencies must be an array');
+    }
+    const dependencyLines = additionalAppDependencies
+      .map(function (dependency) {
+        if (/[\s']/.test(dependency)) {
+          throw new Error(`Disallowed characters in an additional dependency: ${dependency}`);
```

Comment:
we probably would like to show which chars are not acceptable in the error message